clients/drcachesim/tracer/func_trace.cpp

```diff
@@ -0,0 +1,238 @@
+        append_entry(drcontext, TRACE_MARKER_TYPE_FUNC_ARG, arg_i);
+    }
+}
+
+// NOTE: try to avoid invoking any code that could be traced by func_post_hook
+//       (e.g., stl, libc, etc.)
+static void
+func_post_hook(void *wrapcxt, void *user_data)
+{
+    void *drcontext = drwrap_get_drcontext(wrapcxt);
+    if (drcontext == NULL)
+        return;
+
+    size_t idx = (size_t)user_data;
+    func_metadata_t *f = (func_metadata_t *)drvector_get_entry(&funcs, idx);
+    uintptr_t retval = (uintptr_t)drwrap_get_retval(wrapcxt);
+    append_entry(drcontext, TRACE_MARKER_TYPE_FUNC_ID, (uintptr_t)f->id);
+    append_entry(drcontext, TRACE_MARKER_TYPE_FUNC_RETVAL, retval);
+}
+
+static void
+instru_funcs_module_load(void *drcontext, const module_data_t *mod, bool loaded)
+{
+    size_t offset;
+    drsym_error_t result;
+    app_pc func_pc;
+
+    if (drcontext == NULL || mod == NULL)
+        return;
+    dr_log(NULL, DR_LOG_ALL, 1,
+           "instru_funcs_module_load start=%p, mod->full_path=%s\n",
+           mod->start, mod->full_path);
+
+    for (size_t i = 0; i < funcs.entries; i++) {
+        func_metadata_t *f = (func_metadata_t *)drvector_get_entry(&funcs, i);
+        result = drsym_lookup_symbol(mod->full_path, f->name, &offset,
+                                     DRSYM_DEMANGLE);
+        dr_log(NULL, DR_LOG_ALL, 1,
+               "func_name=%s, id=%d, arg_num=%d, result=%d, offset=%d\n",
+               f->name, f->id, f->arg_num, result, offset);
+
+        if (result == DRSYM_SUCCESS) {
+            func_pc = mod->start + offset;
+            drwrap_wrap_ex(func_pc, func_pre_hook, func_post_hook, (void *)i, 0);
+        }
+    }
+}
+
+static std::vector<std::string>
+split_by(std::string s, std::string sep)
+{
+    size_t pos;
+    std::vector<std::string> vec;
+    do {
+        pos = s.find(sep);
+        vec.push_back(s.substr(0, pos));
+        s.erase(0, pos + sep.length());
+    }
+    while (pos != std::string::npos);
+    return vec;
+}
+
+static bool
+id_existed(int id)
+{
+    for (uint i = 0; i < funcs.entries; i++) {
+        func_metadata_t *f = (func_metadata_t *)drvector_get_entry(&funcs, i);
+        if (id == f->id)
+            return true;
+    }
+    return false;
+}
+
+DR_EXPORT
+bool
+func_trace_init(func_trace_append_entry_t append_entry_)
+{
+    std::string value, sep;
+
+    if (dr_atomic_add32_return_sum(&func_trace_init_count, 1) > 1)
+        return true;
+    if (op_record_function.get_value().empty())
+        return true;
+
+    if (!drvector_init(&funcs, 32, false, free_func_entry)) {
+        DR_ASSERT(false);
+        goto failed;
+    }
+
+    append_entry = append_entry_;
+    value = op_record_function.get_value();
+    sep = op_record_function.get_value_separator();
+    for (auto &single_op_value : split_by(value, sep)) {
+        auto items = split_by(single_op_value, PATTERN_SEPARATOR);
+        if (items.size() != 3) {
+            DR_ASSERT(false);
```

Comment:
This is user input so instead of an assert this should be a usage error with a message to the user: the FATAL() macro: `FATAL("Usage error: -record_function was not passed a triplet")` or sthg.

Reply:
Done